solutions/usaco-833.mdx

```diff
@@ -329,189 +328,255 @@ this solution if you find it unfamiliar and come back later.
+		StringTokenizer initial = new StringTokenizer(read.readLine());
+		int N = Integer.parseInt(initial.nextToken());
+		String[] cows = new String[]{initial.nextToken(), initial.nextToken()};
 
+		Map<String, String> edge = new HashMap<>();
 		for (int i = 0; i < N; i++) {
-			st = new StringTokenizer(in.readLine());
-			String mother = st.nextToken(), child = st.nextToken();
-			// store the mother-child relationship as a directed graph
-			// in an adjacency list, where the arrow points to the mother
-			// e.g.: child -> mother -> grandmother
-			// note that the outdegree of each node is maximal 1 in this question
-			adjacencyList.put(child, mother);
-		}
-		in.close();
-
-		// distance from node X and node Y to other nodes
-		// e.g.: xDistance.get(node) = the distance between node and x
-		Map<String, Integer> xDistance = new HashMap<String, Integer>(), 
-			yDistance = new HashMap<String, Integer>();
-		xDistance.put(X, 0);
-		yDistance.put(Y, 0);
-
-		// calculate the distances between X and its ancestors
-		String current = X;
-		while (adjacencyList.containsKey(current)) {
-			String mother = adjacencyList.get(current);
-			xDistance.put(mother, xDistance.get(current) + 1);
-			current = mother;
+			StringTokenizer relation = new StringTokenizer(read.readLine());
+			String v = relation.nextToken();
+			String u = relation.nextToken();
+			// If v is a mother of u, then there is an edge from node u to node v.
+			edge.put(u, v);
 		}
 
-		// cf. above, this time for Y
-		current = Y;
-		while (adjacencyList.containsKey(current)) {
-			String mother = adjacencyList.get(current);
-			yDistance.put(mother, yDistance.get(current) + 1);
-			current = mother;
+		Map<String, Integer>[] dists = new HashMap[]{
+			new HashMap<>(), new HashMap<>()
+		};
+		for (int c = 0; c < 2; c++) {
+			// BFS to find the ancestors of both cows and the distance to them
+			String at = cows[c];
+			dists[c].put(at, 0);
+			while (true) {
+				if (edge.containsKey(at)) {
+					String next = edge.get(at);
+					dists[c].put(next, dists[c].get(at) + 1);
+					at = next;
+				} else {
+					break;
+				}
+			}
 		}
 
-		// minimum distance between shared ancestors of X and Y
-		int minXDist = Integer.MAX_VALUE, minYDist = Integer.MAX_VALUE;
 		String ancestor = null;
-		// iterate through the ancestors of X
-		// and test if Y also has this node as ancestor
-		for (var pair : xDistance.entrySet()) {
-			if (yDistance.containsKey(pair.getKey())
-					// we update the minimum distance if there is an ancestor
-					// with less distance to X or Y than current minimum
-					&& (minXDist > pair.getValue() || minYDist > yDistance.get(pair.getKey()))) {
-				ancestor = pair.getKey();
-				minXDist = pair.getValue();
-				minYDist = yDistance.get(pair.getKey());
+		int dist0 = Integer.MAX_VALUE;  // distance from cow 0 to the common ancestor
+		int dist1 = Integer.MAX_VALUE;  // and the distance from cow 1
+		for (var d0 : dists[0].entrySet()) {
```

Comment:
are you looping through cows? why is it `d`?

Reply:
? i'm looping through all the nodes that are ancestors of cow 0

Reply:
then name it `cow` or somethin

Reply:
calling it `u`, `v`, or `adj` should be ok